# Dailiy Summary:


* QC gff files, if everything is ok, re-run HTseq count, and if everything goes smoothely PCA and DE analysis
	* of course it wasn't, typo in PBS script. Fighting with typos. Can't deal with this right now. 
	* something is very wrong with HM43 annotation:
		* don't know if it's get-homologs, original gff, or me?
		

	
* By next Wednesday want to compare results to Ali, and have something smart to say

* Start looking at eQTL tutorial, tutorial I sent to Katie, and co-clustering analysis
	* Need to read more: basics: need SNP profile + expression data
	* Looking at Ali's variant calling pipeline - do I have to re-align all the reads to 'referecnce' genome then create VCF file?

* side question: what is the spread in expression of ribosomal genes
* how to create a coverage vector over the whole genome
	* maybe start with a part of genome: two parts actually next to origin and next to terminus, or even select specific genes as rerpesentative to start off with

* split cells
* count plates from yesterday

	* same trend as yesterday, which biologically does not make sense. Now starting to wonder if for whatever reason things on the *right* side of the plate always have a growth defect. Next week switch physical location of cft and Deut, and the rest
	
* read ods on E.coli growing on EA
	* looks good

	


