# Daily Summary:

## Cell culture experiment:

* if this one and a repeat show **nothing** will abandon cell culture all together
* Today: infect with CFT, Deut, Deut + R, deutR + R, MOI 25, 24 and 48 hr infections. Don't leave cells to dry out!!!! Wash 3x. Everything with EA. 
* Protocol as always: infect for 1 hr, gent @ 100 ug/ml for 1 hr, gent @ 10 ug/ml for 24 or 48 h. Lyse and plate -1 dilution (tomorrow and Thursday)

* Another way to see if eutR complementation is functional - look for growth on EA alone in minimal media on the bench (MOPS), give it ~48 hr
	* Going to take reading at 24h and 48 h
	* [EA/B12] [5mm/150nM] MOPS: 200 ul + either 2 or 20 ul of o/n culture 

	
## Calculate RPKMs for the count data
	
* for these log2 transform, calculate mean, log2FC
* pathway analysis on these
* see if this matches Ali's results + now want to 
* look at appropriate filtering steps to be taken
	
	### What actually happened:

	* Started working on RPKM.py script, realized that for unknown mysetrious reason _gh.gff files are empty...
	* Then accidentaly deleted folder containing all the homolog information
	* had issues accessing the data on Ali's flux
	* Starting over: might as well put everything together in one pbs script
	* Tired
	* Eventually re-created all the gffs, updated RPKM.py file
		* calculated RPKMs from counts: loop through files `for i in $ALIGN/*.bam; do NUM=$(ls $i|sed -e s/[^0-9]//g); echo "HM"$NUM"_gh.gff"; done`
		* combinded RPKMs in ipython with pandas
		
	
## Still to be done before meeting tomorrow:	
* Take a quick look at co-expression patterns, and phylogenetic signal tools 
* Look up UPEC transcriptome paper, put notes together for meeting tomorrow

