            
####  PBS preamble

#PBS -N gff_file 

#PBS -M annasint@umich.edu
#PBS -m bea
#PBS -j oe

#PBS -l nodes=1:ppn=4,pmem=4gb,walltime=12:00:00
#PBS -V

#PBS -A hmobley_fluxod
#PBS -l qos=flux
#PBS -q fluxod

####  End PBS preamble
####  Commands follow this line

# This script requires python3

if [ -s "$PBS_NODEFILE" ] ; then
    echo "Running on"
    cat $PBS_NODEFILE
fi

if [ -d "$PBS_O_WORKDIR" ] ; then
    cd $PBS_O_WORKDIR
    echo "Running from $PBS_O_WORKDIR"
fi
DATA=/scratch/hmobley_fluxod/annasint/HUTI_RNAseq/Data/get-homologs
SN=/scratch/hmobley_fluxod/annasint/HUTI_RNAseq/Data
CODE=/scratch/hmobley_fluxod/annasint/HUTI_RNAseq/Analysis/lib/get-homologs
#cd $DATA
#for i in *fna
#do FILE=$(ls $i)
#echo ">>"${FILE%.*} > ${FILE%.*}"_name.fna"
#cat $FILE >> ${FILE%.*}"_name.fna"
#rm $FILE
#done

#cat *.fna > upec_homologs.fna

cd $SN
python $CODE/get_gffs_from_homolgs.py upec_homologs.fna upec_homologs.gff 1
mv upec_homologs.gff $SN/Annotations
cd $SN/Annotations
while read p; do grep "^HM"$p"_1" upec_homologs.gff > "HM"$p"_gh.gff"; done < $SN/strain.numbers


