            
####  PBS preamble

#PBS -N htseq-count 

#PBS -M annasint@umich.edu
#PBS -m bea
#PBS -j oe

#PBS -l nodes=1:ppn=4,pmem=4gb,walltime=12:00:00
#PBS -V

#PBS -A hmobley_fluxod
#PBS -l qos=flux
#PBS -q fluxod

####  End PBS preamble
####  Commands follow this line

# This script requires python2-anaconda, samtools

if [ -s "$PBS_NODEFILE" ] ; then
    echo "Running on"
    cat $PBS_NODEFILE
fi

if [ -d "$PBS_O_WORKDIR" ] ; then
    cd $PBS_O_WORKDIR
    echo "Running from $PBS_O_WORKDIR"
fi
DATA=/scratch/hmobley_fluxod/annasint/HUTI_RNAseq/Data
ALIGN=/scratch/hmobley_fluxod/annasint/HUTI_RNAseq/Data/Alignments/BAM_by_Position_New_Samples
ANNOT=/scratch/hmobley_fluxod/annasint/HUTI_RNAseq/Data/Annotations/get-homologs-gff
ANALYSIS=/scratch/hmobley_fluxod/annasint/HUTI_RNAseq/Analysis/core-genome-analysis/2017-01-30-htseq-counts

#htseq-count -f bam -r pos -m union -t CDS -s yes $ALIGN/UR17_SE_aln_sort.bam $ANNOT/HM17_gh.gff > HM17.out.1

for i in $ALIGN/*.bam
do
NUM=$(ls $i|sed -e s/[^0-9]//g)
NAM=$(ls $i|cut -d "/" -f9|cut -d "_" -f1)
htseq-count -f bam -r pos -m union -t CDS -s yes $i $ANNOT/"HM"$NUM"_gh_final.gff"> $ANALYSIS/$NAM
done

