---
title: "Non-core analysis"
author: "ASintsova"
date: "4/26/2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
dat = "/Users/annasintsova/git_repos/HUTI-RNAseq/analysis/non-core/data/"
fig = "/Users/annasintsova/git_repos/HUTI-RNAseq/analysis/non-core/figures/"
library(DESeq2)
library(ggplot2)
library(AnnotationDbi)
library(pathview)
library(gage)
library(org.EcK12.eg.db)
library(readr)

```

## Goal: look for differential gene expression of genes not present in all the genomes

1. Generate datasets with counts for genes that are present in 12, 11... 5 genomes. Does not matter what specific genomes they belonged to. Run DESeq2 on each of those datasets, run PCA 
2. Combine differentiall expressed gene list, and do a pathway analysis


## 1. Generate datasets with counts for genes that are present in 12, 11... 5 genomes. Does not matter what specific genomes they belonged to

```{r generating_gene_sets}
#Load in DESeq counts
counts <- read.csv(paste0(dat, "2017-02-01-combined_counts.csv"), row.names = 1)
strain_names <- c("UR1", "UTI1", "UR2", "UTI2", "UR3", "UTI3", "UR4", "UTI4", "UR5", "UTI5", "UR6", "UTI6", "UR7", "UTI7",
  "UR8", "UTI8", "UR9", "UTI9", "UR10", "UTI10", "UR11", "UTI11", "UR12", "UTI12")

exp_info <- read.csv(paste0(dat, "non-core-exp-info.csv"), row.names = 1)
cmb.sig.genes <- data.frame()


####

#Trying to make this into an actual script

genome.num <- 5:12


for (i in genome.num){
  
  
  num.missing <- 26 - i*2
  print(num.missing)
  all.count <- counts[rowSums(is.na(counts)) == num.missing,]
  cts <- matrix(NA, ncol = i*2, nrow = nrow(all.count))
  rownames(cts)<-rownames(all.count)
  colnames(cts)<- strain_names[1:ncol(cts)]
  for (j in 1:nrow(all.count)){
    mod <- all.count[j,]
    
    cts[j, ]<- mod[!is.na(mod)]
    
  }

  info <- exp_info[rownames(exp_info) %in% colnames(cts),,drop = FALSE]
  
  cts.dds <- DESeqDataSetFromMatrix(countData = cts, colData = info,
                                    design = ~MEDIA)
  
  cts.dds.transformed <- rlog(cts.dds, blind=FALSE)
  pcaData <- plotPCA(cts.dds.transformed, intgroup = c("MEDIA"), returnData = TRUE)
  percentVar <- round(100*attr(pcaData, "percentVar"))
  g <- ggplot(data = pcaData, aes(x=PC1, y=PC2, color = MEDIA)) + 
    geom_point(size =3) + xlab(paste0("PC1: ", percentVar[1], "% var")) +
    ylab(paste0("PC2: ", percentVar[2], "% variance")) +
    coord_fixed()
  plot(g)
  
  
  cts.de <- DESeq(cts.dds, quiet = TRUE)
  
  cts.sign <- subset(results(cts.de, lfcThreshold = 1), padj < 0.05)
  cts.sign <- cts.sign[order(cts.sign$log2FoldChange, decreasing = TRUE),]
  
  
  cmb.sig.genes <- rbind(cmb.sig.genes, cts.sign)
  
  
}

cmb.sig.genes <- cmb.sig.genes[order(cmb.sig.genes$log2FoldChange, decreasing = TRUE),]

```

## 2 Combine differentiall expressed gene list, and do a pathway analysis



```{r get_edited_data, echo=FALSE}
gene_info <- read.table(paste0(dat, "non_core_gene_info.txt"), header = TRUE, row.names = 1, sep = "\t")
non.core <- merge(cmb.sig.genes, gene_info[, c(2,4)], by = "row.names")
rownames(non.core)<- non.core[,1]
non.core <- non.core[,2:9]
non.core <- non.core[order(non.core$log2FoldChange, decreasing = TRUE),]
head(non.core,10)
```

```{r path_analysis}
kegg.ec <- kegg.gsets(species = "eco", id.type = "eco")
kg.ec <- kegg.ec$kg.sets[kegg.ec$sigmet.idx]
p.a <- non.core[,2]
names(p.a) <- non.core[,8]

path.analysis <- gage(p.a, gsets = kg.ec)
print('upregulated')
head(path.analysis$greater[,1:5], 4)
print('downregulated')
head(path.analysis$less[,1:5], 4)

```


* Combining core and non-core differentially expressed genes


```{r combined_path_analysis}
core <- read.csv("/Users/annasintsova/git_repos/HUTI-RNAseq/analysis/DE/2017-04-03-pathway-analysis-revisited/data/DE_genes_edited.csv",
                 row.names = 1)

all.genes <- rbind(core[,c(2,9)], non.core[,c(2,8)])

p.a <- all.genes[,1]
names(p.a) <- all.genes[,2]

path.analysis <- gage(p.a, gsets = kg.ec)
print('upregulated')
head(path.analysis$greater[,1:5], 10)
print('downregulated')
head(path.analysis$less[,1:5], 10)
```


* Compare this to core alone:


```{r core_pa}
core.p.a <- core[,2]
names(core.p.a)<-core[,9]

path.analysis <- gage(core.p.a, gsets = kg.ec)
print('upregulated')
head(path.analysis$greater[,1:5], 10)
print('downregulated')
head(path.analysis$less[,1:5], 10)
```

