{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "import sys\n",
    "\n",
    "sys.path.append('/Users/annasintsova/git_repos/HUTI-RNAseq/code/methods')\n",
    "import keggAPI\n",
    "import helpers\n",
    "out_dir = \"/Users/annasintsova/git_repos/HUTI-RNAseq/results/meta_analysis\"\n",
    "#file_path = \"/Users/annasintsova/git_repos/HUTI-RNAseq/data/sarguru_files/SRR1278952_edited_counts.txt\"\n",
    "file_path = \"/Users/annasintsova/git_repos/HUTI-RNAseq/results/meta_analysis/hm56_to_mg1655_counts.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>count</th>\n",
       "      <th>bnums</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ID=gene0;Dbxref=EcoGene:EG11277;Name=thrL;gbke...</td>\n",
       "      <td>266</td>\n",
       "      <td>b0001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ID=gene1;Dbxref=EcoGene:EG10998;Name=thrA;gbke...</td>\n",
       "      <td>1474</td>\n",
       "      <td>b0002</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ID=gene2;Dbxref=EcoGene:EG10999;Name=thrB;gbke...</td>\n",
       "      <td>1069</td>\n",
       "      <td>b0003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ID=gene3;Dbxref=EcoGene:EG11000;Name=thrC;gbke...</td>\n",
       "      <td>620</td>\n",
       "      <td>b0004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ID=gene4;Dbxref=EcoGene:EG14384;Name=yaaX;gbke...</td>\n",
       "      <td>32</td>\n",
       "      <td>b0005</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                gene  count  bnums\n",
       "1  ID=gene0;Dbxref=EcoGene:EG11277;Name=thrL;gbke...    266  b0001\n",
       "3  ID=gene1;Dbxref=EcoGene:EG10998;Name=thrA;gbke...   1474  b0002\n",
       "5  ID=gene2;Dbxref=EcoGene:EG10999;Name=thrB;gbke...   1069  b0003\n",
       "7  ID=gene3;Dbxref=EcoGene:EG11000;Name=thrC;gbke...    620  b0004\n",
       "9  ID=gene4;Dbxref=EcoGene:EG14384;Name=yaaX;gbke...     32  b0005"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table = pd.read_csv(file_path, sep = \"\\t\", header=None)[[8, 9]]\n",
    "table.columns = [\"gene\", \"count\"]\n",
    "table = table[table.gene.str.contains(\"locus_tag=\")]\n",
    "bnums = [x.split(\"locus_tag=\")[1].split(\";\")[0] for x in table.gene]\n",
    "table['bnums'] = bnums\n",
    "table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(4518, 1)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts = table[[\"bnums\", \"count\"]].set_index(\"bnums\")\n",
    "counts.head()\n",
    "counts.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "22 57 421\n"
     ]
    }
   ],
   "source": [
    "rRNA_genes = keggAPI.searchKEGGGenome(\"eco\", \"ribosomal RNA\", out_dir)\n",
    "with open (\"/Users/annasintsova/git_repos/HUTI-RNAseq/\"\\\n",
    "           \"results/resource_allocation/\"\\\n",
    "           \"2018-04-16_eco_ribosomal_subunit.kegg.results\", \"r\") as fh:\n",
    "    ribosome_genes = [line.split()[0].rstrip().split(\":\")[1].strip() for line in fh]\n",
    "    \n",
    "with open (\"/Users/annasintsova/git_repos/HUTI-RNAseq/\"\\\n",
    "           \"results/resource_allocation/\"\\\n",
    "           \"2018-04-13_crp_activator_info.tab\", \"r\") as fh:\n",
    "    crp_genes = [line.split()[0].rstrip() for line in fh]\n",
    "print(len(rRNA_genes), len(ribosome_genes), len(crp_genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "count    12212465\n",
      "dtype: int64\n",
      "count    28.259496\n",
      "dtype: float64\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "count    9.678734\n",
       "dtype: float64"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rRNA_genes\n",
    "counts = counts.drop(rRNA_genes, axis=0)\n",
    "counts.shape\n",
    "total_reads = counts.sum(axis=0)\n",
    "print(total_reads)\n",
    "ribosome_counts = counts.loc[ribosome_genes].sum(axis=0)\n",
    "print(ribosome_counts/total_reads*100)\n",
    "crp_genes_edited = [c for c in crp_genes if c in counts.index]\n",
    "#print(\"Missing {} crp genes\".format(len(crp_genes) - len(crp_genes_edited)))\n",
    "crp_counts = counts.loc[crp_genes_edited].sum(axis=0)\n",
    "crp_counts/total_reads*100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "ribosome_counts = df_treat.loc[ribosome_genes].sum(axis=0)\n",
    "ribosome_counts\n",
    "print(len(ribosome_counts) == 21)\n",
    "crp_genes_edited = [c for c in crp_genes if c in df_treat.index]\n",
    "print(\"Missing {} crp genes\".format(len(crp_genes) - len(crp_genes_edited)))\n",
    "crp_counts = df_treat.loc[crp_genes_edited].sum(axis=0)\n",
    "print(len(crp_counts) == 21)\n",
    "\n",
    "totals =  pd.DataFrame({'Total': total_reads,'R':ribosome_counts,\n",
    "                        'C': crp_counts})\n",
    "totals[\"C%\"] = round(totals[\"C\"]/totals[\"Total\"]*100,2)\n",
    "totals[\"R%\"] = round(totals[\"R\"]/totals[\"Total\"]*100,2)\n",
    "#totals['condition'] = [c.split(\"_\")[1] for c in totals.index]\n",
    "totals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>bnums</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>b0023</th>\n",
       "      <td>56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0169</th>\n",
       "      <td>325</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b0911</th>\n",
       "      <td>982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b1089</th>\n",
       "      <td>54</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b1716</th>\n",
       "      <td>56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b1717</th>\n",
       "      <td>95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2185</th>\n",
       "      <td>204</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2597</th>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2606</th>\n",
       "      <td>40</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b2609</th>\n",
       "      <td>61</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3065</th>\n",
       "      <td>70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3165</th>\n",
       "      <td>85</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3180</th>\n",
       "      <td>64</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3185</th>\n",
       "      <td>175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3186</th>\n",
       "      <td>320</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3230</th>\n",
       "      <td>188</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3231</th>\n",
       "      <td>164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3259</th>\n",
       "      <td>122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3294</th>\n",
       "      <td>247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3296</th>\n",
       "      <td>895</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3297</th>\n",
       "      <td>368</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3298</th>\n",
       "      <td>261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3299</th>\n",
       "      <td>105</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3301</th>\n",
       "      <td>114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3302</th>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3303</th>\n",
       "      <td>129</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3304</th>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3305</th>\n",
       "      <td>173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3306</th>\n",
       "      <td>121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3307</th>\n",
       "      <td>148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3308</th>\n",
       "      <td>226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3309</th>\n",
       "      <td>154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3310</th>\n",
       "      <td>280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3311</th>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3312</th>\n",
       "      <td>67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3313</th>\n",
       "      <td>207</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3314</th>\n",
       "      <td>347</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3315</th>\n",
       "      <td>276</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3316</th>\n",
       "      <td>503</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3317</th>\n",
       "      <td>510</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3318</th>\n",
       "      <td>443</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3319</th>\n",
       "      <td>753</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3320</th>\n",
       "      <td>490</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3321</th>\n",
       "      <td>353</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3341</th>\n",
       "      <td>452</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3342</th>\n",
       "      <td>233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3636</th>\n",
       "      <td>98</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3637</th>\n",
       "      <td>157</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3703</th>\n",
       "      <td>108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3936</th>\n",
       "      <td>181</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3983</th>\n",
       "      <td>201</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3984</th>\n",
       "      <td>454</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3985</th>\n",
       "      <td>541</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b3986</th>\n",
       "      <td>567</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4200</th>\n",
       "      <td>198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4202</th>\n",
       "      <td>103</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>b4203</th>\n",
       "      <td>126</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       count\n",
       "bnums       \n",
       "b0023     56\n",
       "b0169    325\n",
       "b0911    982\n",
       "b1089     54\n",
       "b1716     56\n",
       "b1717     95\n",
       "b2185    204\n",
       "b2597     24\n",
       "b2606     40\n",
       "b2609     61\n",
       "b3065     70\n",
       "b3165     85\n",
       "b3180     64\n",
       "b3185    175\n",
       "b3186    320\n",
       "b3230    188\n",
       "b3231    164\n",
       "b3259    122\n",
       "b3294    247\n",
       "b3296    895\n",
       "b3297    368\n",
       "b3298    261\n",
       "b3299    105\n",
       "b3301    114\n",
       "b3302     42\n",
       "b3303    129\n",
       "b3304    119\n",
       "b3305    173\n",
       "b3306    121\n",
       "b3307    148\n",
       "b3308    226\n",
       "b3309    154\n",
       "b3310    280\n",
       "b3311     42\n",
       "b3312     67\n",
       "b3313    207\n",
       "b3314    347\n",
       "b3315    276\n",
       "b3316    503\n",
       "b3317    510\n",
       "b3318    443\n",
       "b3319    753\n",
       "b3320    490\n",
       "b3321    353\n",
       "b3341    452\n",
       "b3342    233\n",
       "b3636     98\n",
       "b3637    157\n",
       "b3703    108\n",
       "b3936    181\n",
       "b3983    201\n",
       "b3984    454\n",
       "b3985    541\n",
       "b3986    567\n",
       "b4200    198\n",
       "b4202    103\n",
       "b4203    126"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts.loc[ribosome_genes]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
