---
title: "Genetic Components of PCA"
author: "ASintsova"
date: "4/20/2017"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(stats)
library(devtools)
#install_github("ggbiplot", "vqv")
library(ggbiplot)
library(pheatmap)
data ="/Users/annasintsova/git_repos/HUTI-RNAseq/analysis/evolutionaryModels/2017-04-12-PC2-loadings/data/"
fig = "/Users/annasintsova/git_repos/HUTI-RNAseq/analysis/evolutionaryModels/2017-04-12-PC2-loadings/figures/"

```

## Goals

1. Re-normalize 'normalized' data and re-do the PCA to see if there are any drastic changes

2. Using PCA from re-normalized data determine PC2 (phylogroup) component loadings for combined data, urine alone, uti alone

3. Run phylogenetic tests on genes identified to see if they show 'phylogenetic signal'

4. Come up with further plan of action 



## 1. Re-normalize 'normalized' data and re-do the PCA to see if there are any drastic changes

* From Steve:

```{r inverse_normalization}

invnorm = function(x) { qnorm((rank(x,na.last="keep",ties.method="random")-0.5)/sum(!is.na(x))); }

```

* From the internet

```{r alt_inv_norm}
my.invnorm = function(x)
{
 res = rank(x)
 res = qnorm(res/(length(res)+0.5))
 return(res)
}

```


* Import normalized counts & expreiment_info: count distribution is very far from normal


```{r load_data, echo = FALSE}
counts <- read.csv(paste0(data, "normalized_counts_core.csv"), row.names = 1)
info <- read.csv("/Users/annasintsova/git_repos/HUTI-RNAseq/analysis/DE/data/Patient_meta_info.csv",
                row.names = 1)
hist(counts$UR1) # very far from normal
```


* If we do log2 transform, looks much better


```{r log2, echo = FALSE}
hist(log2(counts$UR1 +1)) 

```

* Steve's inverse normalization


```{r inverse_norm_hist}

 hist(invnorm(counts$UR1)) 

```

* And using the other formula:

```{r}

 hist(my.invnorm(counts$UR1))

```


* Inverse normalizing all the data

```{r inverse_norm_counts, echo = FALSE}

inv.norm.c <- matrix(NA, ncol = 26, nrow = 2502, byrow = FALSE)
 rownames(inv.norm.c)<- rownames(counts)
 colnames(inv.norm.c) <- colnames(counts)

 for (i in 1:ncol(counts)){
   x <- counts[,i]
   norm.x <- invnorm(x)
   inv.norm.c[,i] <- norm.x

 }
 head(inv.norm.c)
```


* Performing PCA analysis on these inverse normalized counts. Plotting all samples, label by media




```{r setting_up_pca, echo = FALSE}
t.norm.counts <- t(inv.norm.c)
t.norm.counts <- merge(t.norm.counts, info[, c(2,4,5)], by ="row.names")
t.norm.counts <- t.norm.counts[,2:2506]
rownames(t.norm.counts)<- colnames(inv.norm.c)
media <- t.norm.counts[, 2503]
his <- t.norm.counts[,2504]
psg <- t.norm.counts[,2505]
all.pca <- prcomp(t.norm.counts[,1:2502])

g <- ggbiplot(all.pca, obs.scale = 1, var.scale = 1,
             groups = media, ellipse = TRUE, circle = FALSE,
             var.axes = FALSE)
g <- g + scale_color_discrete(name = '')
g <- g + theme(legend.direction = 'horizontal',
                             legend.position = 'top')
g


```


* Plotting all samples by phylogroup


```{r pca_by_phylogroup}

g <- ggbiplot(all.pca, obs.scale = 1, var.scale = 1,
              groups = psg, ellipse = TRUE, circle = FALSE,
                var.axes = FALSE)
g <- g + scale_color_discrete(name = '')
g <- g + theme(legend.direction = 'horizontal',
                legend.position = 'top')
g

```


* Urine samples alone


```{r urine_pg, echo = FALSE}
odd <- seq(1, 26, 2)
t.urine <- t.norm.counts[odd, ]
u.psg <- t.urine[, 2505]
u.his <- t.urine[, 2504]
u.pca <- prcomp(t.urine[,1:2502])

#plotting by psg
g <- ggbiplot(u.pca, obs.scale = 1, var.scale = 1,
              groups = u.psg, ellipse = TRUE, circle = FALSE,
                var.axes = FALSE)
g <- g + scale_color_discrete(name = '')
g <- g + theme(legend.direction = 'horizontal',
                legend.position = 'top')
g

```

* Just for to see what would a random clustering look, plot by 'history'


```{r urine_history, echo = FALSE}
#plotting by history
g <- ggbiplot(u.pca, obs.scale = 1, var.scale = 1,
              groups = u.his, ellipse = TRUE, circle = FALSE,
                var.axes = FALSE)
g <- g + scale_color_discrete(name = '')
g <- g + theme(legend.direction = 'horizontal',
                legend.position = 'top')
g

```

* UTI samples, by phylogroup

```{r uti_phylogroup, echo = FALSE}
even <- seq(2, 26, 2)
t.uti <- t.norm.counts[even, ]
uti.psg <- t.uti[, 2505]
uti.his <- t.uti[, 2504]
uti.pca <- prcomp(t.uti[,1:2502])

#plotting by psg
g <- ggbiplot(uti.pca, obs.scale = 1, var.scale = 1,
              groups = uti.psg, ellipse = TRUE, circle = FALSE,
                var.axes = FALSE)
g <- g + scale_color_discrete(name = '')
g <- g + theme(legend.direction = 'horizontal',
                legend.position = 'top')
g

```


* UTI samples by history


```{r}
g <- ggbiplot(uti.pca, obs.scale = 1, var.scale = 1,
              groups = uti.his, ellipse = TRUE, circle = FALSE,
                var.axes = FALSE)
g <- g + scale_color_discrete(name = '')
g <- g + theme(legend.direction = 'horizontal',
                legend.position = 'top')

g
```


## 2. Using PCA from re-normalized data determine PC2 (phylogroup) component loadings for combined data, urine alone, uti alone

* Determine PC2 loadings (a.k.a. rotation)


```{r pc_loadings}
annot = info[, c("MEDIA", "PRED_PHYLO")]
pca.data <- prcomp(t(inv.norm.c))
pc2.loadings <- pca.data$rotation[,2][order(pca.data$rotation[,2], decreasing = TRUE)][1:10]#PC2
names <- names(pc2.loadings)
load.genes <- inv.norm.c[names,]
head(names, 10)

```

* Visualize these 

```{r pc2_heatmap, echo = FALSE}

pheatmap(load.genes, annotation_col = annot, cellheight = 10, cellwidth = 10, cluster_rows = FALSE)#,


```

* Also look at PC1 loadings


```{r pc1_loadings, echo = FALSE}

pc1.loadings <- pca.data$rotation[,1][order(pca.data$rotation[,1], decreasing = TRUE)][1:10]#PC1
names <- names(pc1.loadings)
load.genes <- inv.norm.c[names,]

pheatmap(load.genes, annotation_col = annot, cellheight = 10, cellwidth = 10, cluster_rows = FALSE)
          
```


* Look at the location of these genes on B vs B1 genome, use HM14 and HM56

```{r pc2_hm14}


hm14.pc2.locs <- c(3831641, 1816299, 933957, 2035632, 936157, 2615289, 2615678, 2616476, 2617043, 2617428,3570210, 937776, 4160540, 1465521 )
names(hm14.pc2.locs) <- c("ori","argS","wzc1", "wzc2", "wzb", "rplS", "trmD", "rimM", "rpsP", "ffh", "bcsB", "hypo1", "hypo2", "termi")
hist(hm14.pc2.locs, 100, xlim = c(0, 4400000))
abline(v = 1465521, col = 'blue')
abline(v = 3831641, col = 'red')
abline(v = 2035632, col = "orange")
abline(v = 4160540, col = "orange" )

```


* HM56


```{r hm56_pc2_genes}

all.locs <- read.table(paste0(data, "pc2_ed.gff"), sep = "\t")
get_id <- function(x){
  gi <- unlist(strsplit(unlist(strsplit(x, c(";")))[1], "="))[2]
  return (gi)
}

g.ids<- sapply(as.character(all.locs[,3]), get_id)
all.locs[,3]<- g.ids
colnames(all.locs)<- c("genome", "location", "gene")


hm56.pc2 <- all.locs[all.locs$genome == "HM56_1",]
hm56.pc2.locs <- hm56.pc2$location
names(hm56.pc2.locs)<- hm56.pc2$gene

hm56.pc2.locs


hist(hm56.pc2.locs, 100, xlim = c(0, 4000000))
abline(v = 3177816, col = 'blue')
abline(v = 764444, col = 'red')



```

* Now do the same thing only using urine data and visualize 



```{r pc2_urine_alone}
annot = info[, c("MEDIA", "PRED_PHYLO")]
pca.data.urine <- prcomp(t.urine[,1:2502])
pc2.loadings.urine <- pca.data.urine$rotation[,2][order(pca.data.urine$rotation[,2], decreasing = TRUE)][1:5]#PC2
names.urine <- names(pc2.loadings.urine)
load.genes.urine <-inv.norm.c[names.urine, odd]
head(names.urine, 10)

pheatmap(load.genes.urine, annotation_col = annot, cellheight = 10, cellwidth = 10, cluster_rows = FALSE)#,

```


* Same for only UTI samples...


```{r pc2_uti_alone}
annot = info[, c("MEDIA", "PRED_PHYLO")]
pca.data.uti <- prcomp(t.uti[,1:2502])
pc2.loadings.uti <- pca.data.uti$rotation[,2][order(pca.data.uti$rotation[,2], decreasing = TRUE)][1:20]#PC2
names.uti <- names(pc2.loadings.uti)
load.genes.uti <-inv.norm.c[names.uti, even]
head(names.uti, 10)

pheatmap(load.genes.uti, annotation_col = annot, cellheight = 10, cellwidth = 10, cluster_rows = FALSE)#,

```


## 3. Run phylogenetic tests on genes identified to see if they show 'phylogenetic signal'



* Have to remember how I did this before...



