# Daily Summary:

* One thing Evan wanted me to look at was what the core genes look like for Sarguru's strains

	* Same issue as before: how do I pullout all the core genes? 
	* One option would be to blast all the genes identified by get homologs agains each of Sarguru's genomes and get their locs and count read for that

	
* Continuing with phylogenetic signal analysis

	* Easy: look at one gene for which there was very strong signal
		
		* quick look in IGV
		* align upstream/downstream sequences
		* regulon DB
		
	* Identified a problem right away: it appears that the problem is as always with annotation: HM14, HM7, and HM3 have 2 genes (wzc and etk) that were given the same gene_id under get-homologs

	
* Trying to think through the data:

	* Concerns: what data should I be putthing through? right now working with normalized counts
	
	* Looking at urine data alone - possibly enrichment of biofilm expression genes in B1 phylogroup? could be interesting

* Ran oriC on all of our genomes trying to locate origin of replication

	* idea will be for each gene to be able to calculate  distance from the origin, see if there are any patterns


	
* A brief look into what is coming up appears to indicate enrichment in response regulators/ 2 component systems and membrane proteins - potentially could be interesting

	
	

	
	